### PR TITLE
A Seller's ability to mark product/item availability,delete and update

### PR DESCRIPTION
**Pull request by suave990**
### What does this PR do?

Providing access to a seller to mark his/her products as available or not , delete the product, and update the product from his/her collection

### Description of Task to be completed?
Why?
Products should be availed for buyers to purchase

As a seller I should be able to mark a product as available or unavailable manually in order to make it available for the buyers on not.

Reasons can be: expiry date, low on stock, etc

As a Seller
I should be able to delete a specific item in my colletion
Such that I can remove items I no longer deal in/were errornously added

Acceptance criteria

Given a delete request to delete a specific item in a given seller's collection
Scenario 1
When the requesting user/token is a seller and the target item does not exist in the particular seller's collection and the request body is validated
Then:

A json object with a status: 4xx, error: the appropriate error message should be sent
Scenario 2
When the requesting user/token is a seller and the target item already exist in the particular seller's collection and the request body is validated
Then:

The item should be deleted
A json object containing status: the appropriate code, message: appropriate response message, and item: an object of the deleted item should returned
Scenario 3
When the request body fails any validation checks or the token is not of the role seller
Then:

The appropriate 4XX response message should be sent to the requesting client describing the validation error
Dev notes:

All inputs should be validated for sanity"

As a Seller
I should be able to update a specific item in my collection
Such that items always have the right/required status, details and specifications

Acceptance criteria
Given a put/patch request to update all a particular item of a given seller
Scenario 1
When the requesting user/token is a seller and the target item does not exist in the particular seller's collection and the request body is validated
Then:

A json object with a status: 4xx, error: the appropriate error message should be sent
Scenario 2
When the requesting user/token is a seller and the target item already exist in the particular seller's collection and the request body is validated
Then:
The item should be updated
A json object containing status: the appropriate code, message: appropriate response message, and item: an object of the updated item should returned
Scenario 3
When the request body fails any validation checks or the token is not of the role seller
Then:

The appropriate 4XX response message should be sent to the requesting client describing the validation error
Dev notes:

All inputs should be validated for sanity"

### How should this be manually tested?
using swagger documentation
### Any background context you want to provide?
n/a
Any other information you can give about the task.
n/a
### What are the relevant pivotal tracker stories?

Provide a relevant link to the pivotal tracker.
https://www.pivotaltracker.com/n/projects/2628461/stories/184347100
https://www.pivotaltracker.com/n/projects/2628461/stories/184347097
https://www.pivotaltracker.com/n/projects/2628461/stories/184347096
### Screenshots (if appropriate)

n/a
### Questions:

# Type of change
feature to controllers and services
Please check the options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or Enhancement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code is formatted correctly
- [x] My changes doesn't have any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by Ntarecedrick — DISMISSED** [review]

**Review by NG-Solennel — DISMISSED** [review]

File: src/routes/api/product.route.js [hunk 1]
```diff
@@ -24,10 +24,32 @@ route.get(
...
   extractToken,
   checkRole(["admin"]),
   IsProductExist,
-  ProductWishesController.getWishProductByProduct);
+  ProductWishesController.getWishProductByProduct,
+);
+route.post(
+  "/:id",
+  extractToken,
+  checkRole(["seller"]),
```
Comment (Eli250) [inline src/routes/api/product.route.js:37]: You need to add a middleware to check if the seller is a product owner before updating any product. Do it for all product requests (PUT, DELETE)

File: src/routes/api/product.route.js [hunk 2]
```diff
@@ -24,10 +24,32 @@ route.get(
...
-route.get("/:id/productWishes",
+route.get(
+  "/:id/productWishes",
   extractToken,
   checkRole(["admin"]),
   IsProductExist,
-  ProductWishesController.getWishProductByProduct);
+  ProductWishesController.getWishProductByProduct,
+);
+route.post(
```
Comment (Eli250) [inline src/routes/api/product.route.js]: This is PATCH not a POST request

**Review by Eli250 — CHANGES_REQUESTED** [review]

File: src/routes/api/product.route.js (hunk 1 above)
Comment (suave990) [reply src/routes/api/product.route.js:37]: Ok am going to add that middleware to all the 3

**Review by istafrica-dennw — CHANGES_REQUESTED** [review]
I was reviewing this piece of code but saw some irregularities in logic
you marking product as available
you check if the note is "available" (after already marking)
then you send a response to client based on condition in (2)
This  logic is not real.  if you want to check teh note you need to do it so before sending command to the database to update.
More, could you also make is unavailable, a seller might have other reasons he does not want to avail it.
Questions to ask yourself
are you validating the body? (availaible vs unavailable note: only these two options should be allowed)

**Review by istafrica-dennw — CHANGES_REQUESTED** [review]
response.note
what is response? isn't this the product?
I would recommend to rename it to product

and note? shouldn't this be availabllity?

File: src/app.js [hunk 3]
```diff
@@ -39,7 +39,7 @@ try {
     console.log(`server running on port ${port} `);
   });
 } catch (error) {
-  console.log(error);
+  // console.log(error);
```
Comment (iradukundairenee) [inline src/app.js]: please remove this comment

File: src/middlewares/checkOwner.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+const { Product } = require("../database/models");
```
Comment (iradukundairenee) [inline src/middlewares/checkOwner.js]: please use this  instead import { Product } from "../database/models";

File: src/services/productService.js [hunk 5]
```diff
@@ -274,5 +274,34 @@ class ProductServices {
...
+    const images = req.files.map((img) => img.path);
+    await Product.update(
+      { price, name, bonus, expiryDate, categoryId, quantity, images },
+      { where: { id: req.params.id } },
+    );
+  }
+
+  static async deleteItem(id) {
+    await Product.destroy({ where: { id } });
+  }
```
Comment (iradukundairenee) [inline src/services/productService.js]: @suave990  The code does not include any error handling for cases where the database operations fail. Adding appropriate error handling can make the code more robust and prevent unexpected crashes.

File: src/services/productService.js [hunk 6]
```diff
@@ -274,5 +274,34 @@ class ProductServices {
...
```
Comment (iradukundairenee) [inline src/services/productService.js]: also The naming conventions used in the code should be consistent. For example, in markAvailable method, id is passed as a parameter, while in updateItem and deleteItem methods, id is passed as an argument. It is better to choose one style and stick to it throughout the code.

**Review by iradukundairenee — CHANGES_REQUESTED** [review]

File: src/services/productService.js (hunk 6 above)
Comment (suave990) [reply src/services/productService.js]: yes! I undestand the need for consistency

File: src/routes/api/product.route.js (hunk 1 above)
Comment (suave990) [reply src/routes/api/product.route.js:37]: done

**Review by Eli250 — APPROVED** [review]

**Pull request merged**